### PR TITLE
refactor: setup project for TypeScript

### PR DESCRIPTION
**Pull request by santimirandarp**
Not sure about these:

* There is this one extra change respect to the migration guide, see the only "chore" in commits.

* The auto generated type is less strict than the current type SimpleLinearRegression, maybe there is clean solution, otherwise I'd propose to change `"prepack": "npm run tsc"` to 

```
"prepack": "npm run tsc && cp regression.d.ts lib/index.d.ts"
```

_but this is not included yet._

-------------

 @targos @lpatiny

### CODE REVIEW

**Comment by targos** [comment]
> There is this one extra change respect to the migration guide, see the only "chore" in commits.

This is unrelated to TS migration. This line was actually needed before but we haven't touched the project in a while so it wasn't done.

> The auto generated type is less strict than the current type SimpleLinearRegression, maybe there is clean solution, otherwise I'd propose to change "prepack": "npm run tsc" to

Can you explain how it's less strict? In any case, we should delete the old typings.

File: .github/workflows/nodejs.yml [hunk 1]
```diff
@@ -11,4 +11,5 @@ jobs:
     # Documentation: https://github.com/zakodium/workflows#nodejs-ci
     uses: zakodium/workflows/.github/workflows/nodejs.yml@nodejs-v1
     with:
+      lint-check-types: true
       node-version-matrix: '[12, 14, 16]'
```
Comment (targos) [inline .github/workflows/nodejs.yml:15]: Please remove this line. We now inherit the default version matrix in most projects

File: package.json [hunk 2]
```diff
@@ -2,25 +2,30 @@
...
-  "types": "regression-simple-linear.d.ts",
+  "main": "./lib/index.js",
+  "module": "./lib-esm/index.js",
+  "types": "./lib/index.d.ts",
   "sideEffects": false,
   "files": [
-    "regression-simple-linear.d.ts",
     "lib",
+    "lib-esm",
+    "regression-simple-linear.d.ts",
```
Comment (targos) [inline package.json:12]: to be removed

File: src/__tests__/test.js [hunk 3]
```diff
@@ -7,7 +7,9 @@ describe('Simple Linear Regression', () => {
 
     const regression = new SLR(inputs, outputs);
 
+    // @ts-expect-error coefficients possibly undefined
     expect(regression.slope).toBe(regression.coefficients[1]);
+    // @ts-expect-error coefficients[1] possibly undefined
     expect(regression.intercept).toBe(regression.coefficients[0]);
```
Comment (targos) [inline src/__tests__/test.js:13]: In what case can coefficients be undefined?

Comment (santimirandarp) [reply src/__tests__/test.js:13]: I think this is in the case that `coefficients` key doesn't exist

Comment (santimirandarp) [reply src/__tests__/test.js:13]: I've done the typing on an extra branch, and the test comments can be removed in the next PR

**Comment by targos** [comment]
I feel like we are not on the same page. You are free to do all the additional changes and refactorings you think can improve the project, even if they are not mentioned in the TS migration guide.

In this case, the typings issue is because we have absolutely no jsdoc comments in the code, so TS has no choice but to use `any`.
I understand if you don't want to redo the typings correctly now, but it will have to be done before the next release.

**Comment by santimirandarp** [comment]
It's not a problem, I did it already, I just thought you wanted changes in several stages. They'll be included in a few minutes in this PR.

File: package.json [hunk 4]
```diff
@@ -2,25 +2,29 @@
...
-    "prepack": "npm run compile",
+    "prepack": "npm run tsc",
     "prettier": "prettier --check src",
     "prettier-write": "prettier --write src",
-    "test": "npm run test-coverage && npm run eslint && npm run prettier",
-    "test-only": "jest",
-    "test-coverage": "jest --coverage"
+    "test": "npm run test-coverage && npm run eslint && npm run prettier && npm run check-types",
+    "test-only": "vitest --run",
+    "test-coverage": "vitest --run --coverage",
```
Comment (targos) [inline package.json:24]: You can remove the `test-coverage` script and add the coverage option to `test-only`.

File: src/__tests__/test.ts [hunk 5]
```diff
@@ -119,6 +119,7 @@ describe('Simple Linear Regression', () => {
   });
 
   it('wrong model', () => {
+    //@ts-expect-error types not fully compatible yet
     expect(() => SLR.load({})).toThrow(/^not a SLR model$/);
```
Comment (targos) [inline src/__tests__/test.ts:123]: Here we really expect an error, types should not allow this.

File: src/index.ts [hunk 6]
```diff
@@ -3,16 +3,25 @@ import BaseRegression, {
...
 
+type JsonType = ReturnType<SimpleLinearRegression['toJSON']>;
+
 export default class SimpleLinearRegression extends BaseRegression {
-  constructor(x, y) {
+  slope!: number;
+  intercept!: number;
+  coefficients!: number[];
+
+  constructor(x: true, y: JsonType);
```
Comment (targos) [inline src/index.ts:13]: this constructor is internal, we shouldn't expose it in the types

**Comment by santimirandarp** [comment]
[GH Pages Deployment sample](https://santimirandarp.github.io/regression-simple-linear/
)

**Comment by targos** [comment]
Changes look good. Did you intend to keep it draft?

**Comment by santimirandarp** [comment]
> Changes look good. Did you intend to keep it draft?

😃 

I'm drafting a [little migrator following the guide](https://github.com/santimirandarp/ts-migration-helper), to run from codespaces. 

I will move to PR as the eslint error seems unrelated ?

**Comment by targos** [comment]
I've never seen this ESLint error, it's suspicious. Rerunning the job...

**Comment by santimirandarp** [comment]
> I've never seen this ESLint error, it's suspicious. Rerunning the job...

Yes, it does not appear locally, I couldn't find out why.

**Pull request merged**